src/items.rs

```diff
@@ -592,31 +592,95 @@ impl<'a> FmtVisitor<'a> {
         self.buffer.push_str(&generics_str);
 
         self.last_pos = body_start;
-        self.block_indent = self.block_indent.block_indent(self.config);
-        for (i, f) in enum_def.variants.iter().enumerate() {
-            let next_span_start: BytePos = if i == enum_def.variants.len() - 1 {
-                span.hi
-            } else {
-                enum_def.variants[i + 1].span.lo
-            };
 
-            self.visit_variant(f, i == enum_def.variants.len() - 1, next_span_start);
+        // Need to preserve comments if no variants are present
+        if enum_def.variants.is_empty() {
+            self.format_missing_with_indent(span.hi - BytePos(1));
+        } else {
+            self.buffer.push_str("\n");
+            let indentation = self.block_indent.block_indent(self.config).to_string(self.config);
+
+            self.buffer.push_str(&indentation);
+            self.block_indent = self.block_indent.block_indent(self.config);
+
+            if let Some(body_str) = self.format_variant_list(enum_def,
+                                                             body_start,
+                                                             span.hi - BytePos(1)) {
+                self.buffer.push_str(&body_str);
+            }
+            self.block_indent = self.block_indent.block_unindent(self.config);
+
         }
-        self.block_indent = self.block_indent.block_unindent(self.config);
 
-        self.format_missing_with_indent(span.hi - BytePos(1));
+        self.buffer.push_str(&self.block_indent.to_string(self.config));
         self.buffer.push_str("}");
+        self.last_pos = span.hi;
+    }
+
+    // Format the body of an enum definition
+    fn format_variant_list(&self,
+                           enum_def: &ast::EnumDef,
+                           body_lo: BytePos,
+                           body_hi: BytePos)
+                           -> Option<String> {
+
```

Comment:
Nit: the blank lines after opening braces are not really in line with the rest of the code.